<a href="https://colab.research.google.com/github/pernilletemmerud/DCF-PT/blob/main/Notebook2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Pandas and Numpy


Python has several built-in types of data structures such as lists and dictionaries, but none of them allows you to efficiently do computations on large lists or tables of data. The Numpy and Pandas libraries fill this gap by providing 1) specific data structures on which one can efficiently compute and 2) functions to perform the computations on these structures.

Numpy is the core scientific computing library on which virtually all other scientific Python packages are building (except for deep learning libraries). This includes notably Scipy, a package offering numerical routines for optimization, regression etc. that will be used in the last series of lectures.

Pandas, which itself builds on top of Numpy, is the core data science library on which higher level libraries (e.g. for plotting) are building. 

Here we briefly present the main data structure offered by Pandas, the **DataFrame** and then show how the data in DataFrames are actually **arrays**, the main data structure of Numpy. This allows us in the next notebooks to explore Numpy before coming back at the end of the course to Pandas.

Let's first import both Pandas and Numpy. Both have very commonly used abbreviations ```np``` and ```pd``` that you should use as well to simplify your life. Note that most sub-modules are directly accessible and you won't have to import specific ones as migth be the case for other libraries.

In [ ]:
import pandas as pd
import numpy as np

## Importing data

Just like any other library, Pandas offers functionality through the *dot-notation* (as e.g. in ```math.cos()```). When surveying usage of Pandas in places like GitHub, it turns out that the most used function is the one allowing one to import a CSV (comma-separated values). Indeed while we could import tabular data with the ```read``` function, this becomes quickly cumbersome for complex data and Pandas takes care of all details for us. 


### read_*** functions

As an example we import here the file available [here](https://github.com/KingaS03/Introduction-to-Python/blob/master/Homework/country_vaccinations.xls). If we look at the file we see that even though it has an .xls extension, it is in fact just a CSV file. Pandas has many importing functions and we use here ```read_csv```. You can see the list of all available importers by typing ```pd.read_``` and seeing options from autocomplete.

We can either download the file and indicate the file path or directly *read* it from the web (*Note that on GitHub you need to use the [raw](https://raw.githubusercontent.com/KingaS03/Introduction-to-Python/master/Homework/country_vaccinations.xls) version of the file to download it*)

```
url = 'https://raw.githubusercontent.com/KingaS03/Introduction-to-Python/master/Homework/country_vaccinations.xls'
vaccine = pd.read_csv(url)
```

Let's first see what this object is:

```
type(vaccine)
```

We see that we are dealing with a DataFrame, the two-dimensional data structure offered by Pandas. If we just execute a cell with that variable we can display the first lines:

```
vaccine
```

We see that we are dealing with tabular data of various types: we have text (e.g. countries), dates and numbers. Just like in a tabular sheet, each column has a label and each line has an index (column in bold on the left), so that every element can be located with these *coordinates*.

## Methods attached to dataframes

We have seen before that every variable in Python has a series of functions attached to it (methods). For example if we have a text variable (string) we can e.g. split it at a given place:

```
my_string = 'This is a sentence with spaces.'
my_string.split(' ')
```

The same logic applies to Pandas DataFrames and in general to any new variable that is created by a given package: they all come with a series of methods attached to them.

For example if we only want to display the fist 5 lines of the DataFrame, we can use the ```head``` method:

vaccine.head(5)

We will learn later how to extract statistics, but just as an example of functions that exsit, you can for example get a basic statistical description of a DataFrame using the ```describe``` method:

vaccine.describe()

We see that in this particular case, the returned object is a DataFrame as well!

## Accessing columns

If we want to work with only a given column from a dataframe, we can extract it. Just like when we want to extract an element from a regular list, e.g. ```my_list[3]```, for Pandas we also use the square parenthesis (brackets) but this time with the name of the column:

```
some_column = vaccine['daily_vaccinations']
some_column
```

As we have a single columns here, we are not dealing with a DataFrame anymore but with a Series:

type(some_column)

We won't explore much the Series object on its own in this course, but know that you can for example create a DataFrame from scratch by combining multiple Series.

## What is underlying Pandas

We see above that ```some_column``` is composed of and index (0, 1, 2 etc) and the corresponding values (NaN, 64.0, 64.0 etc.). If we want to do mathematics with those values, we don't really need the index and a list of numbers would be sufficient. We can obtain such a list by asking for the ```values``` of the series:

```actual_values = some_column.values
actual_values
```

We see that the output is not just a simple Python list. It is in fact called an ```array```. If we ask for the type of this object, we get:

type(actual_values)

The lists of values contained in Pandas DataFrame are in fact Numpy arrays. A large part of the computing logic in Pandas is inspired from Numpy, therefore we now make a detour to Numpy before coming back to DataFrames.

## Exercise


1. In the Data folder on Ilias there is the data file 4ms_10 av_30 s_Absorbance_10-32-04-868.txt. Upload this file to a folder on your google drive. If you don't have a google account, make it now. Look at the file by clicking on it. Try to make sense of the data with the help of the metadata file also in the Ilias folder. 

2. Mount the file to this your notebook (example code below). The path and the file name must match the ones on your google drive, of course.


In [12]:
from google.colab import drive
drive.mount("/content/drive", force_remount=True)
path = '/content/drive/MyDrive/DataScienceFundamentals/4ms_10 av_30 s_Absorbance_10-32-04-868.txt' # To be adapted

Mounted at /content/drive


3. Study and play with the code below. Try to understand based on the API description (https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html). Hint: Use the header argument.

In [13]:
import pandas as pd
df = pd.read_csv(path, delimiter = '\t', header=13)

4. Study this dataset a bit with the df.head(), df.info() and df.describe() methods. The available metadata is [here](https://docs.google.com/document/d/1L6m4OFiCKSHlHlr4_4TFKLKHcJfIkiMhHtD3tcsxpD8/edit?usp=sharing)

In [14]:
df.head()

Unnamed: 0     Unnamed: 1  ...  779.9194  780.0381
0  2018-10-25 10:32:04.789000  1540456324789  ... -0.039100 -0.035545
1  2018-10-25 10:32:04.878000  1540456324878  ... -0.040210 -0.038165
2  2018-10-25 10:32:04.962000  1540456324962  ... -0.038686 -0.036903
3  2018-10-25 10:32:05.046000  1540456325046  ... -0.038203 -0.036822
4  2018-10-25 10:32:05.129000  1540456325129  ... -0.038082 -0.035900

[5 rows x 3650 columns]

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 355 entries, 0 to 354
Columns: 3650 entries, Unnamed: 0 to 780.0381
dtypes: float64(3648), int64(1), object(1)
memory usage: 9.9+ MB


In [16]:
df.describe()

Unnamed: 1    316.6137    316.7492  ...    779.8008    779.9194    780.0381
count  3.550000e+02  355.000000  355.000000  ...  355.000000  355.000000  355.000000
mean   1.540456e+12    0.409207    0.409207  ...   -0.036634   -0.037724   -0.035371
std    8.676718e+03    0.327439    0.327439  ...    0.001624    0.001545    0.001536
min    1.540456e+12   -0.243277   -0.243277  ...   -0.039905   -0.040929   -0.038231
25%    1.540456e+12    0.103420    0.103420  ...   -0.037541   -0.038611   -0.036165
50%    1.540456e+12    0.330891    0.330891  ...   -0.036708   -0.037786   -0.035425
75%    1.540456e+12    0.844202    0.844202  ...   -0.035855   -0.036984   -0.034661
max    1.540456e+12    0.871232    0.871232  ...   -0.015445   -0.016981   -0.014757

[8 rows x 3649 columns]

How big is the dataset? How many rows, how many columns, how space in the memory? --> 8 rows and 3649 columns


The data are spectral lines from some chemistry experiment with some other information in the first four columns. We will come back to this dataset later.  

5. Plot some columns.

In [18]:
#df[5].hist()

6. Check out other pandas read methods here: https://pandas.pydata.org/docs/reference/io.html

### Notes lecture 2

In [ ]:
#multiplying regular lists
alist = [1, 2, 3]
alist*3 

[1, 2, 3, 1, 2, 3, 1, 2, 3]

In [ ]:
import numpy as np

#maths with arrays
alist = np.array([1, 2, 3])
print(alist*3)
print(alist-2)
print(alist/2)

[3 6 9]
[-1  0  1]
[0.5 1.  1.5]


In [ ]:
#removing item from list

#clear() - remove all items
#pop() - remove an item by idex and get its value
#remove() - remove an item by value
#del() - remove an item by index or slice

alist = ['a', 1, 'hi', 999]
print(alist.pop(1))
alist




1


['a', 'hi', 999]